sensor/mcp9808/mcp9808.go

```diff
@@ -318,65 +308,77 @@ type Temperature struct {
 	AboveCritical, AboveUpper, BelowLower bool
 }
 
-// AmbientTemp reads the current sensor value along with the flags denoting what boundaries the
-// current temperature exceeds.
-func (d *MCP9808) AmbientTemp() (*Temperature, error) {
-	temp, err := d.Bus.ReadWordFromReg(address, regAmbientTemp)
-	if err != nil {
-		return nil, err
-	}
-
-	tempC := temp &^ 0xF000
-
-	wholeNum := float64(tempC&^0xF) / 16.0
-	fraction := tempC &^ 0xFF0
-
-	tempResult := &Temperature{
-		AboveCritical: !(temp&(1<<15) == 0),
-		AboveUpper:    !(temp&(1<<14) == 0),
-		BelowLower:    !(temp&(1<<13) == 0)}
-	tempResult.CelsiusDeg = wholeNum + (float64(fraction) / 10000.0)
-	if !(temp&(1<<12) == 0) { // read sign bit
-		tempResult.CelsiusDeg *= -1
-	}
-	return tempResult, nil
-}
-
-// readTempC reads from the ambient temperature register and returns the current temperature value in celsius
+// readTempC reads from the reg temperature register and returns the current temperature value in celsius
 func (d *MCP9808) readTempC(reg byte) (float64, error) {
 	temp, err := d.Bus.ReadWordFromReg(address, reg)
 	if err != nil {
 		return 0, err
 	}
 
-	log.Printf("\nTA vs TCrit: %v\nTA vs TUpper: %v\nTA vs TLower: %v\n", !(temp&(1<<15) == 0), !(temp&(1<<14) == 0), !(temp&(1<<13) == 0))
+	return convertWordToTempC(temp), nil
+}
 
-	wholeTempC := temp &^ 0xF000
+func convertWordToTempC(temp uint16) float64 {
+	wholeNum := float64(temp&0xFF0) / 16.0
+	fraction := float64(temp&0xF) * .0625
 
-	wholeNum := float64(wholeTempC&^0xF) / 16.0
-	fraction := wholeTempC &^ 0xFF0
+	tempRead := wholeNum + fraction
 
-	tempRead := wholeNum + (float64(fraction) / 10000.0)
 	if !(temp&(1<<12) == 0) { // read sign bit
 		tempRead *= -1
 	}
-	return tempRead, nil
+	return tempRead
 }
 
 func (d *MCP9808) setTemp(reg byte, newTemp float64) error {
```

Comment:
Which method are you referring to here?  The diff you've selected is inside of the convertWordToTempC function but the line you have in the comment looks like you're referring to the setTemp function.  I've watched the values back and forth as I've set them and both of those functions work as intended (they return the right values).  The setTemp has the switch statement because each temperature that you set can be set to .25 precision.  Your version appears to round to the nearest whole number which is fine but the sensor allows for more precision.  It also doesn't account for the sign bit